### PR TITLE
Add individual sorting of item children

### PR DESCRIPTION
**Pull request by apfelbox**
| Q             | A
| ------------- | --------------------------------------------------------------------- |
| Bug fix?      | no                                                                |
| New feature?  | yes                    |
| Improvement?  | no     |
| BC breaks?    | no                                                                |
| Deprecations? | no     |
| Docs PR       | https://github.com/Becklyn-Studios/docs/pull/32                                   |

Now the items can be sorted individually.

### CODE REVIEW

File: src/Item/MenuItem.php [hunk 1]
```diff
@@ -735,49 +772,13 @@ public function resolveTree (string $currentClass = "current", string $ancestorC
             $this->addListItemClass($ancestorClass);
         }
 
-        // sort by priority
-        $this->children = $this->sort($this->children);
+        // sort children
+        $this->children = MenuItemSorter::sort($this->sort, $this->children);
```
Comment (keichinger) [inline src/Item/MenuItem.php:776]: I'm not too sure about this static dependency here as I generally dislike static method calls inside a plain PHP object like this one. Can't we pass in an item sorter whenever we want to resolve the tree? The caller could then decide which sorting method should be used by simply passing in a different one. That way applications could easily change the sorting implementation, if necessary - although, to be fair, I think we already cover a lot of use cases with the current sorter.

So the `$option["sort"]` could be as it's replaced by a dedicated sorter. This method signature would then look like `public function resolveTree (string $currentClass = "current", string $ancestorClass = "ancestor", int $level = 0, `?MenuItemSorterInterface $itemSorter = null) : bool`

File: tests/Item/MenuItem/MenuItemSortTest.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+use Becklyn\Menu\Exception\InvalidSortMethodException;
+use Becklyn\Menu\Item\MenuItem;
+use Becklyn\Menu\Sorter\MenuItemSorter;
+use PHPUnit\Framework\TestCase;
+
+class MenuItemSortTest extends TestCase
+{
+    private function buildMenu (string $sort) : MenuItem
+    {
+        $root = new MenuItem(null, ["sort" => $sort]);
```
Comment (keichinger) [inline tests/Item/MenuItem/MenuItemSortTest.php:14]: Can we also get a test-case with some special chars numbers and one with the same letter in capital and lower case?

**Review by keichinger — COMMENTED** [review]
If you think that we should keep the current implementation as it'd turn out to be an unnecessary overkill to remove the static dependency, I'm fine with it. Implememntation wise this PR looks good and is ready to be merged, as soon as we have a added a few more test menu entries :)

File: tests/Item/MenuItem/MenuItemSortTest.php (hunk 2 above)
Comment (apfelbox) [reply tests/Item/MenuItem/MenuItemSortTest.php:14]: Not sure about what you mean by „special chars numbers“.

Capital letters are already included, the important thing to test (= test category) is the case insensitivity, not the actual letters (= would be no new test category).

File: src/Item/MenuItem.php (hunk 1 above)
Comment (apfelbox) [reply src/Item/MenuItem.php:776]: The point here is that the sorter should be a simple function without any dependencies.  We could in-line it, but that just bloats the menu item. We could make it non-static, but that doesn’t give us anything except worse performance and GC pressure. We could make it a real function, but then autoloading is sort-of broken. 

Passing it in doesn’t work, as the resolveTree call is recursive and every item can have its own sorting (that is the point of this PR). 

So passing a sorter in doesn’t work, we could pass a registry of sorters in. 

But...

a) that is brutally over-engineered 😕
b) that part is quite performance-sensitive and that would just kill all of it. Keep in mind that static calls are super fast in PHP.

**Review by keichinger — APPROVED** [review]

**Pull request merged**